### ISSUE TITLE
test.test_banana: eliminate duplicate code and fix wrong assertion message

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7665]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7665|
|Type|enhancement|
|Created|2014-10-04 08:06:02Z|
|Branch|https://github.com/twisted/twisted/tree/banana-remove-duplication-7665|

This also fixes a future problem with porting to Python 3:
StringIO.truncate(0) implicitly does seek(0) first
but io.bytesIO.truncate(0) does not

Attachments:

* [7665.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d9/8d942504f83368f072d9aaa383c9eb09c08b4910/300004e9cf30b36f0284f8c06c392302eac34c4a.patch) (5733 bytes) - added by wolfgang61 on 2014-10-04 08:08:53Z - 
* [7665b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d9/8d942504f83368f072d9aaa383c9eb09c08b4910/d12f4837992b01e7bf2d8eeb4aec9ec7b3686714.patch) (5830 bytes) - added by wolfgang61 on 2014-10-04 21:31:15Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7665 7665
type__enhancement enhancement
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__ 
branch__branches_banana_remove_duplication_7665 branches/banana-remove-duplication-7665
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__spread spread
time__1412409962912524 1412409962912524
changetime__1412522600482816 1412522600482816
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

  1. twisted.spread is part of "Twisted Core".  News fragments go into `twisted/topfiles`.
  1. Use fully qualified names in news fragments.
  1. The change to `testPartial` is incorrect.  The test is exercising the behavior when the data is not all received at once.  If you want to add a docstring to this test method explaining this, that would be fine.
  1. `twisted.test.proto_helpers.StringTransport` is somewhat more preferred than `twisted.internet.protocol.FileWrapper` these days.  It might be more beneficial to switch the tests to using `StringTransport` rather than preparing them to be switched to `io.BytesIO`.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wolfgang61 set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_1):
>   1. The change to `testPartial` is incorrect.  The test is exercising the behavior when the data is not all received at once.

def feed() does just that: Feeding the data byte per byte. Now I added a docstring to testPartial() and feed(). Or did I misunderstand you? I want to re-use feed() because indexing bytes in Python 3 returns int, so this needs porting.

>   1. `twisted.test.proto_helpers.StringTransport` is somewhat more preferred than `twisted.internet.protocol.FileWrapper` these days.  It might be more beneficial to switch the tests to using `StringTransport` rather than preparing them to be switched to `io.BytesIO`.

I'd rather make that a separate ticket. But then - changing to io.BytesIO is rather easy and I want to focus on PY3. Same can be said about the assert statements. I do not introduce more but do not really want to upgrade everything to current coding standard unless the PY3 port profits from that.

The news fragment should be correct with latest patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [43225]) Branching to 'banana-remove-duplication-7665'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [43226]) Apply 7665b.patch 

refs [#7665](https://github.com/twisted/twisted/issues/7665)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [43227]) expand the docstring for `_encoded`

refs [#7665](https://github.com/twisted/twisted/issues/7665)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [43228]) text in misc fragments is ignored

refs [#7665](https://github.com/twisted/twisted/issues/7665)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [43229]) Just go ahead with StringTransport since it only involves changing the encoding helper function.

refs [#7665](https://github.com/twisted/twisted/issues/7665)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> def feed() does just that: Feeding the data byte per byte.

Thanks.  Sorry about that.  I was mistaken.

> I'd rather make that a separate ticket.

That's reasonable fine.  As I was going through the latest patch I decided to just go ahead with that myself - it was easy for me and not much of a change - hopefully it doesn't disrupt your patchset too much.

This looks good to me now.  Once buildbot is green I'll merge.

> The news fragment should be correct with latest patch.

Sorry, I overlooked that this was a misc fragment in my last review (which seems correct to me).  The text in misc fragments is thrown away (that's what makes them misc, they're not worth describing in the NEWS file).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [43232]) Merge banana-remove-duplication-7665

Author: wolfgang61, exarkun
Reviewer: exarkun
Fixes: [#7665](https://github.com/twisted/twisted/issues/7665)

Remove some duplication in the encoding logic in the twisted.spread.banana unit tests.
